File: examples/openplatform-go-demo/crypto/tripledes.go
package crypto

import (
	"bytes"
	"crypto/cipher"
	"crypto/des"
	"encoding/base64"
	"errors"
)

// TDESCipher 3DES/ECB/PKCS5Padding encryptor/decryptor.
type TDESCipher struct {
	block     cipher.Block
	blockSize int
}

func NewTDESCipher(key []byte) (*TDESCipher, error) {
	k := normalize3DESKey(key)
	block, err := des.NewTripleDESCipher(k)
	if err != nil {
		return nil, err
	}
	return &TDESCipher{block: block, blockSize: block.BlockSize()}, nil
}

func (c *TDESCipher) Encrypt(plain []byte) (string, error) {
	padded := pkcs5Pad(plain, c.blockSize)
	out := make([]byte, len(padded))
	for i := 0; i < len(padded); i += c.blockSize {
		c.block.Encrypt(out[i:i+c.blockSize], padded[i:i+c.blockSize])
	}
	return base64.StdEncoding.EncodeToString(out), nil
}

func (c *TDESCipher) Decrypt(cipherB64 string) ([]byte, error) {
	raw, err := base64.StdEncoding.DecodeString(cipherB64)
	if err != nil {
		return nil, err
	}
	if len(raw)%c.blockSize != 0 {
		return nil, errors.New("invalid cipher block size")
	}
	out := make([]byte, len(raw))
	for i := 0; i < len(raw); i += c.blockSize {
		c.block.Decrypt(out[i:i+c.blockSize], raw[i:i+c.blockSize])
	}
	return pkcs5Unpad(out)
}

func normalize3DESKey(key []byte) []byte {
	switch len(key) {
	case 16:
		out := make([]byte, 24)
		copy(out, key)
		copy(out[16:], key[:8])
		return out
	case 24:
		return key
	default:
		if len(key) > 24 {
			return key[:24]
		}
		out := make([]byte, 24)
		copy(out, key)
		return out
	}
}

func pkcs5Pad(data []byte, blockSize int) []byte {
	n := blockSize - len(data)%blockSize
	padding := bytes.Repeat([]byte{byte(n)}, n)
	return append(data, padding...)
}

func pkcs5Unpad(data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, errors.New("empty data")
	}
	n := int(data[len(data)-1])
	if n <= 0 || n > len(data) {
		return nil, errors.New("invalid padding")
	}
	for i := 0; i < n; i++ {
		if data[len(data)-1-i] != byte(n) {
			return nil, errors.New("invalid padding bytes")
		}
	}
	return data[:len(data)-n], nil
}

File: examples/openplatform-python-demo/openplatform/crypto/triple_des.py
"""3DES/ECB/PKCS5Padding，与会话密钥交换后的 finalKey 一致。"""

from __future__ import annotations

import base64
import secrets

from Crypto.Cipher import DES3


def random_digits(length: int) -> str:
    return "".join(str(secrets.randbelow(10)) for _ in range(length))


def normalize_key24(key: bytes) -> bytes:
    if len(key) == 16:
        return key + key[:8]
    if len(key) == 24:
        return key
    if len(key) > 24:
        return key[:24]
    return key.ljust(24, b"\x00")


class TripleDesCipher:
    def __init__(self, session_key: str) -> None:
        self._key = normalize_key24(session_key.encode("utf-8"))

    def encrypt(self, plain: bytes) -> str:
        cipher = DES3.new(self._key, DES3.MODE_ECB)
        padded = _pkcs5_pad(plain, cipher.block_size)
        encrypted = cipher.encrypt(padded)
        return base64.b64encode(encrypted).decode("ascii")

    def decrypt(self, cipher_base64: str) -> bytes:
        cipher = DES3.new(self._key, DES3.MODE_ECB)
        raw = base64.b64decode(cipher_base64)
        padded = cipher.decrypt(raw)
        return _pkcs5_unpad(padded)


def _pkcs5_pad(data: bytes, block_size: int) -> bytes:
    n = block_size - len(data) % block_size
    return data + bytes([n] * n)


def _pkcs5_unpad(data: bytes) -> bytes:
    if not data:
        raise ValueError("empty data")
    n = data[-1]
    if n <= 0 or n > len(data):
        raise ValueError("invalid padding")
    if data[-n:] != bytes([n] * n):
        raise ValueError("invalid padding bytes")
    return data[:-n]
